packages/schema/lib/functional-constraints/searchOrCreateKeys.js

```diff
@@ -59,6 +122,149 @@ const validateSearchOrCreateKeys = (definition) => {
         )
       );
     }
+
+    // Confirm searchOrCreate.update matches a creates.key, if it is defined
+    if (updateKey && !definition.creates[updateKey]) {
+      errors.push(
+        new jsonschema.ValidationError(
+          `must match a "key" from a create (options: ${createKeys})`,
+          searchOrCreateDef,
+          '/SearchOrCreateSchema',
+          `instance.searchOrCreates.${key}.update`,
+          'invalidKey'
+        )
+      );
+    }
+
+    // Confirm searchOrCreate.updateInputFromSearchOutput existing implies searchOrCreate.update is defined
+    if (searchOrCreateDef.updateInputFromSearchOutput && !updateKey) {
+      errors.push(
+        new jsonschema.ValidationError(
+          `requires instance.searchOrCreates.${key}.update to be defined`,
+          searchOrCreateDef,
+          '/SearchOrCreateSchema',
+          `instance.searchOrCreates.${key}.updateInputFromSearchOutput`,
+          'invalid'
+        )
+      );
+    }
+
+    // Confirm searchOrCreate.searchUniqueInputToOutputConstraint existing implies searchOrCreate.update is defined
+    if (searchOrCreateDef.searchUniqueInputToOutputConstraint && !updateKey) {
+      errors.push(
+        new jsonschema.ValidationError(
+          `requires instance.searchOrCreates.${key}.update to be defined`,
+          searchOrCreateDef,
+          '/SearchOrCreateSchema',
+          `instance.searchOrCreates.${key}.searchUniqueInputToOutputConstraint`,
+          'invalid'
+        )
+      );
+    }
+
+    // Confirm searchOrCreate.updateInputFromSearchOutput contains objects with:
+    // keys existing in creates[update].operation.inputFields.key
+    // values existing in searches[search].operation.(outputFields.key|sample keys), if they are defined
+    if (searchOrCreateDef.updateInputFromSearchOutput && updateKey) {
+      // Note that _.each({key: value}) provides the following callback method signature: (value, key) => {}
+      _.each(
+        searchOrCreateDef.updateInputFromSearchOutput,
+        (searchOutputField, updateInputField) => {
+          // Confirm searchOrCreate.updateInputFromSearchOutput's key exists in creates[update].operation.inputFields.key
+          if (_.isEmpty(updateInputKeys)) {
```

Comment:
`_.isEmpty(updateInputKeys)` is constant during the loop. So the code would be tidier and more efficient if we move it out of the loop.

Also, it'd be clearer to include `${updateKey}` in the error message.

```javascript
if (searchOrCreateDef.updateInputFromSearchOutput && updateKey) {
  const updateOptionHint = _.isEmpty(updateInputKeys) ? '(no "key" found in inputFields)' : `(options: 
 ${updateInputKeys})`;

  const searchOptionHint = `(options: ${[...allSearchOutputKeys].join(',')})`

  Object.entries(searchOrCreateDef.updateInputFromSearchOutput).forEach((updateInputField, searchOutputField) => {
    if (!updateInputKeys.includes(updateInputField)) {
      errors.push(
        new jsonschema.ValidationError(
          `must match a "key" from a creates.${updateKey}operation.inputFields ${updateOptionHint}`,
          searchOrCreateDef,
          '/SearchOrCreateSchema',
          `instance.searchOrCreates.${key}.updateInputFromSearchOutput`,
          'invalidKey'
        )
      );
    }
    
    if ((hasSearchOutputFields || hasSearchOutputSample) && !allSearchOutputKeys.has(searchOutputField)) {
      errors.push(
        new jsonschema.ValidationError(
          `must match a "key" from searches.${searchKey}.operation.(outputFields|sample) ${searchOptionHint}`,
          searchOrCreateDef,
          '/SearchOrCreateSchema',
          `instance.searchOrCreates.${key}.updateInputFromSearchOutput`,
          'invalidKey'
        )
      );
    }
  });
}
```

Same goes for the `searchUniqueInputToOutputConstraint` code block below.